New ensemble new learning

neworderofjamie opened the pull request:
No filtered activity-based operation, but otherwise works - Re-adds support for Voja and PES to new-ensemble

mundya committed:
Add new parallel filter

The standard filter can now run in parallel.  The number of cores and
chips it uses can be preset or can be automatically determined from the
expected compute load.

Incorporates suggestions by @neworderofjamie.

mundya committed:
Prune filters with no outgoing connections

Automatically prune out Filter operators which have no outgoing
connections, this is equivalent to removing passthrough Nodes which do
not connect to anything.

mundya committed:
Shared memory ensembles

Use SDRAM to reduce the communication load presented by Ensembles.  Each
Ensemble in a `cluster` writes its spike vector into SDRAM and is then
responsible for decoding a subspace of the entire output space.

Includes suggestions by @neworderofjamie

mundya committed:
Shared memory ensembles

Use SDRAM to reduce the communication load presented by Ensembles.  Each
Ensemble in a `cluster` writes its spike vector into SDRAM and is then
responsible for decoding a subspace of the entire output space.

mundya committed:
Make Region.sizeof_padded use args, kwargs

This makes the API more flexible.

mundya committed:
Optimise passthrough Nodes out of model

Optimises passthrough Nodes out of a model by combining their input and
out signals. Passthrough Nodes can be retained by setting the
`optimize_out` option in the config to False; regardless of this setting
passthrough Nodes associated with Neurons will be removed.

neworderofjamie committed:
Updated ensemble builder to use new semantics

neworderofjamie committed:
Python side of learning rule re-integration
* ConnectionMap.get_signals_from_object builds a DEFAULTdict of lists
* ReceptionParams and TransmissionParams contain reference to learning rule

neworderofjamie committed:
Python side of encoder learning

neworderofjamie committed:
Renamed EnsembleRegions to Regions for terseness - discuss!

neworderofjamie committed:
Renames of c files

neworderofjamie committed:
Tidying of include guards #includes to match new names post-rename

neworderofjamie committed:
flake8

neworderofjamie committed:
Most of c-code integrated - still needs more thought in places
* Filtered activity basically disabled

mundya committed:
Expand partitioning to deal with multiple slices

Creates a new `partition_multiple` tool for partitioning multiple slices
simultaneously; the original `partition` function is rewritten as a
wrapper for this new function.

neworderofjamie committed:
* decode_spike_train doesn't belong in the header - Discuss!
* looping on n suggests neurons which made me more confused than seemed necessary
* PES can't be updated in simulate_neurons

neworderofjamie committed:
Small bits of python learning which failed to merge

neworderofjamie committed:
New C implementation of non-filtered PES

neworderofjamie committed:
c-side of functioning PES
* ensemble.parameters hampers anything approximating 80 character lines - assigned to params in a bunch of places
* Seperate learnt and static decoder and keys regions concatenated when loaded so existing static processing can be employed

neworderofjamie committed:
Python side of working Voja
* Seperate decoders and keys regions for learnt and static decoders
* Static and learnt output space partitioned seperately
* 'Properly' disabled filtered activity mode

neworderofjamie committed:
Correct slicing of PES learning

neworderofjamie committed:
pep8

neworderofjamie committed:
Examples that can be removed from here before merging, but I want to not lose this time :)

neworderofjamie committed:
Small bug fixes in scaled encoder recording - still doesn't actually work

neworderofjamie committed:
Re-added total encoder width parameter - recorded encoders now don't look TOTALLY trashes

neworderofjamie committed:
Re-figured out matplotlib animation and made animated encoder plot

neworderofjamie committed:
write voja-learnt encoder to write location!

neworderofjamie committed:
Profiling
* Tidied up PES function a bit - just takes ensemble pointer rather than numerous parameters
* Added profiler tags
* Renamed skip() nop() as it resulted in some epically confusing errors
* Added PES profiling tags

neworderofjamie committed:
fixed DTCM partitioning

neworderofjamie committed:
how the crap did that compile

neworderofjamie committed:
correct SDRAM usage

neworderofjamie committed:
Voja profiling

neworderofjamie committed:
Fix for numpy 1.8

neworderofjamie committed:
fixup existing tests, add some to cover learnt_slices etc

neworderofjamie committed:
Learning rule connectivity builder tests

neworderofjamie committed:
Caught bug in PESRegion (it's like testing works)

neworderofjamie committed:
beginning of work on PES test - need advice from JH regarding how best to test slicing logic

neworderofjamie committed:
Voja tests

neworderofjamie committed:
Executables

neworderofjamie committed:
Encoder recording test

neworderofjamie committed:
test learning rule probing

neworderofjamie committed:
Extra PES tests

neworderofjamie commented:
I think this has reached the point of review! I may try and add some more tests, but I think the coverage of the learning stuff is reasonable

mundya committed:
Merge pull request #87 from project-rig/new-ensemble

Rewritten ensemble

mundya committed:
Merge pull request #77 from project-rig/parallel-filter2

Parallel passthrough Nodes

neworderofjamie committed:
Spike recording didn't work on clustered ensembles

neworderofjamie committed:
Profiling tweaks
* Re-implemented print-stats in terms of my new favourite function - scipy.stats.binned_statistic
* Also used this rather than mean for writing CSV rows - generally rather more meaningful

mundya committed:
Find more filter parallelism

Decompose the matrix applied by filter operators to reduce DMA and
network bandwidth.

Includes suggestions by @neworderofjamie

neworderofjamie committed:
Reset recording regions between simulation runs

neworderofjamie committed:
Reset recording regions between simulation runs

neworderofjamie committed:
Improved handling of errors during simulation

neworderofjamie committed:
Improved handling of errors during simulation

neworderofjamie committed:
Merge remote-tracking branch 'origin/recording_buffer_reset' into new_ensemble-new_learning

neworderofjamie committed:
Added additional constraint to SMULL instruction - otherwise (codesourcery) assigns sasme registers for output and input :(

neworderofjamie committed:
Added additional constraint to SMULL instruction

- otherwise (codesourcery) assigns same registers for output and input

neworderofjamie committed:
Full encoder width should be used for top level partitioning

mundya committed:
Merge pull request #99 from project-rig/recording_buffer_reset

Recording buffer reset

mundya committed:
Merge pull request #101 from project-rig/smull_fix

SMULL fix

mundya committed:
Fixes ParallelFilterSlice::accepts_signal

Bad merge meant that code wasn't updated to account for transposed
Ensemble Transmission Parameters.

mundya committed:
Fix malloc_or_die

Calls RT_ERROR to kill the core after a malloc fails.

mundya commented:
Dare I ask what you'd like to do with this history?

neworderofjamie committed:
Fix malloc_or_die

Calls RT_ERROR to kill the core after a malloc fails.

neworderofjamie committed:
Thanks @mundya - never actually hits exit

neworderofjamie committed:
Thanks @mundya - never actually hits exit

neworderofjamie committed:
C-side data structures for dma transfer of learnt signals

neworderofjamie committed:
C-side extended DMA state machine

neworderofjamie committed:
Also show errors if application dies during runtime

neworderofjamie committed:
actually run learn_associations on SpiNNaker backend...

neworderofjamie committed:
Python side dma transfer of learnt signals

neworderofjamie committed:
use correct RTE code

neworderofjamie committed:
region-size mismatch

mundya committed:
Merge pull request #100 from project-rig/improved_error_handling

Improved error handling

mundya committed:
Merge pull request #103 from project-rig/fix-parallel-filter

Fixes ParallelFilterSlice::accepts_signal

neworderofjamie committed:
Input filters now get sliced - saves memory and doesn't break Voja implementation

neworderofjamie committed:
ahh shared memory, it works great when you read from it as well as writing

neworderofjamie committed:
whole status is actually quite a lot more useful

neworderofjamie committed:
Mega-optimisation!

mundya committed:
flake8

mundya committed:
Fix a missing commit

mundya committed:
Merge pull request #105 from project-rig/fix-simulation-time-bug2

Fix a missing commit

mundya committed:
More flake8 @neworderofjamie!

mundya committed:
Use new routing table tools

Update to use new routing table APIs from Rig > 0.12.0

Removes a largely useless test.

mundya committed:
Merge pull request #104 from project-rig/fix-malloc-or-die

Fix malloc_or_die

mundya committed:
Merge pull request #106 from project-rig/use-new-rig-api

Use new routing table tools

mundya committed:
Merge pull request #95 from project-rig/remove-passthrough-nodes

Remove passthrough nodes

mundya committed:
Allow overwriting filter widths

mundya committed:
Changes to parameters and exceptions.

Includes suggestions by @tcstewar, thanks!

mundya committed:
Sort passthrough nodes before removal

Remove passthrough Nodes in reverse order of depth in a network.

mundya committed:
Add weight to PTN removal heuristic

Add a weight term to the PTN removal heuristic to account for the
multiple packets per row required when column division of a transform
matrix is applied.

studywolf commented:
So i'm playing around with this code now, and I tried to make two connections into `conn.learning_rule`

``` python
nengo.Connection(target, conn.learning_rule, transform=-1)
nengo.Connection(current, conn.learning_rule)
```

and it looked like the first connection was being bumped off by the second connection. When I combined the two in a separate ensemble beforehand and connected that output to the learning rule things worked fine.

neworderofjamie commented:
I am pretty certain I never thought that was even possible :) I THINK it
would be relatively easy to implement. @mundy - What are your plans for the
integration of this, now worryingly out of date, branch and new_ensemble in
general into master?

On 23 March 2016 at 16:21, Travis DeWolf notifications@github.com wrote:

> So i'm playing around with this code now, and I tried to make two
> connections into conn.learning_rule
> 
> nengo.Connection(target, conn.learning_rule, transform=-1)
> nengo.Connection(current, conn.learning_rule)
> 
> and it looked like the first connection was being bumped off by the second
> connection. When I combined the two in a separate ensemble beforehand and
> connected that output to the learning rule things worked fine.
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> https://github.com/project-rig/nengo_spinnaker/pull/96#issuecomment-200420999

tcstewar commented:
> I am pretty certain I never thought that was even possible :)

:)  Yeah, all our examples only use one input to a learning_rule, but as I've been playing with learning rules more and more I keep doing this multiple connection thing.  Basically since the error population is almost always doing some sum of two different inputs, you can almost always just get rid of the error population and do two different connections into the learning rule.  So we've started getting into the habit of doing that, just for efficiency reasons....  But it is definitely quite surprising.

neworderofjamie commented:
If you could raise an issue and assign it to me, I will add it at some
point in the near future. What are you guys using Nengo SpiNNaker learning
for?

On 23 March 2016 at 16:40, tcstewar notifications@github.com wrote:

> I am pretty certain I never thought that was even possible :)
> 
> :) Yeah, all our examples only use one input to a learning_rule, but as
> I've been playing with learning rules more and more I keep doing this
> multiple connection thing. Basically since the error population is almost
> always doing some sum of two different inputs, you can almost always just
> get rid of the error population and do two different connections into the
> learning rule. So we've started getting into the habit of doing that, just
> for efficiency reasons.... But it is definitely quite surprising.
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> https://github.com/project-rig/nengo_spinnaker/pull/96#issuecomment-200431018

studywolf commented:
Robots! :smile:

studywolf commented:
issue made but i am unable to assign!

mundya commented:
> What are your plans for the integration of this, now worryingly out of date, branch and new_ensemble in general into master?

Unsure. Given how far ~~we've~~ I've let this slip it might be easier to do the merge in person when I'm next in Manchester (like w/b 11th April)?

mundya commented:
@neworderofjamie I'll try to review this tomorrow, hopefully the merge shouldn't be too horrible... I don't think master has changed _that_ much.

neworderofjamie commented:
Awesome, has new_ensemble even made it to master? :)

On 23 March 2016 at 18:00, Andrew Mundy notifications@github.com wrote:

> @neworderofjamie https://github.com/neworderofjamie I'll try to review
> this tomorrow, hopefully the merge shouldn't be too horrible... I don't
> think master has changed _that_ much.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/project-rig/nengo_spinnaker/pull/96#issuecomment-200467979

mundya commented:
> Awesome, has new_ensemble even made it to master? :)

Yes, I think it's been in master for a bit now

mundya committed:
Avoid breaking change to `build_decoders`

Changes to `build_decoders` in nengo/nengo#979 were breaking for
nengo_spinnaker.

This change simply copies an older implementation of `build_decoders`
into nengo_spinnaker - probably not an ideal solution.  Some discussion
as to what API backends should and shouldn't use might be a good idea to
avoid this kind of thing regularly happening.

mundya commented on nengo_spinnaker/builder/ensemble.py:
```diff
@@ -221,3 +284,31 @@ def build_neurons_probe(model, probe):
+            # If learning rule modifies encoders
+            if learning_rule_type.modifies == "encoders":
+                # If post-synaptic object is an ensemble, add probe
+                # to the bject operators lists of local probes
```
"add probe to the operator's list of local probes"

mundya committed:
Merge pull request #110 from project-rig/fix-params

Track Nengo upstream

mundya committed:
Remove silent failure in filter operator

As suggested by @neworderofjamie

mundya committed:
Merge pull request #107 from project-rig/more-filter-parallelism

More filter parallelism

mundya committed:
Transition to Rig 1.0.0

mundya committed:
Partition value probes to reduce packet-fan in

TODO:
 - Modify Python to match changes to C

mundya committed:
fixup! Python side of parallel probes

mundya committed:
Merge pull request #112 from project-rig/rig-1.0.0

Transition to Rig 1.0.0

mundya committed:
Documentation fixes

Thanks @mossblaser

mundya committed:
Specify SpiNNaker machine in simulator init

Allow specification of the SpiNNaker machine when initialising a
Simulator object. A SpiNNaker machine allocated by `spalloc` can be used
with:

```python
from spalloc import Job

with Job(3) as j:
    sim = Simulator(model, hostname=j.hostname)
    ...
```

mundya committed:
Automatically allocate port for Node communication

Allow the operating system to select a free port for communication with
a SpiNNaker machine. This may require some changes to firewall settings.

mundya committed:
Defer booting the SpiNNaker machine

Don't construct a machine controller or boot the SpiNNaker machine until
the netlist has been constructed; this will make automatic provision of
SpiNNaker machines with `spalloc` simpler.

mundya committed:
Allow automatic allocation of a SpiNNaker machine

Add a `use_spalloc` argument to the Simulator. If `use_spalloc` is True
then `spalloc` is used to allocate a SpiNNaker machine to perform the
simulation.

mundya committed:
Improvements to spalloc usage

Thanks @mossblaser

mundya committed:
Increase pre-boot delay for spalloc

mundya committed:
Allow setting of use_spalloc in config file

Adding `use_spalloc: True` or similar to the nengo_spinnaker
configuration file will default nengo_spinnaker to using a
spalloc-allocated machine.

Some slight changes made to logging messages.

mundya committed:
Reuse code more

Thanks for spotting that @neworderofjamie

mundya committed:
Report number of boards used, not number requested

Update the logging information to report the number of boards allocated
by spalloc rather than the number that were requested.

mundya committed:
Changes for larger machines.

Thanks @neworderofjamie and @mossblaser.

neworderofjamie committed:
Update to spinnaker_tools 2.0.0

neworderofjamie committed:
Updates to profile_communication_channel example including fix for #118

neworderofjamie committed:
New executables built with spinnaker tools 2.0.0

mundya committed:
Reduce width of value sink operator

Should reduce packet dropping for particularly large SPA models.

mundya committed:
Merge pull request #116 from project-rig/parallel-probe

Parallel probe

mundya committed:
Merge pull request #117 from project-rig/use-spalloc

Use spalloc to get a SpiNNaker machine

mundya committed:
Merge pull request #111 from project-rig/update-passthrough-node-removal

Improve passthrough Node removal

mundya committed:
Fix LinearFilter parameter packing

This could resize the buffer and resulted in nonsense values anyway.

mundya committed:
Merge pull request #120 from project-rig/fix-linear-filter-packing

Fix LinearFilter parameter packing

mundya committed:
Safely remove passthrough Nodes with no input

Fixes a bug which resulted from attempting to remove a passthrough Node
with no incoming connections.

neworderofjamie committed:
Merge pull request #121 from project-rig/remove-sourceless-ptns

Safely remove passthrough Nodes with no input

neworderofjamie committed:
Merge remote-tracking branch 'origin/master' into new_ensemble-new_learning

Conflicts:
	nengo_spinnaker/binaries/nengo_ensemble.aplx
	nengo_spinnaker/binaries/nengo_ensemble_profiled.aplx
	nengo_spinnaker/binaries/nengo_filter.aplx
	nengo_spinnaker/binaries/nengo_mc_player.aplx
	nengo_spinnaker/binaries/nengo_rx.aplx
	nengo_spinnaker/binaries/nengo_tx.aplx
	nengo_spinnaker/binaries/nengo_value_sink.aplx
	nengo_spinnaker/binaries/nengo_value_source.aplx
	nengo_spinnaker/operators/lif.py
	nengo_spinnaker/regions/__init__.py
	nengo_spinnaker/regions/filters.py
	nengo_spinnaker/regions/recording.py
	nengo_spinnaker/simulator.py
	spinnaker_components/ensemble/Makefile
	spinnaker_components/ensemble/ensemble.c
	spinnaker_components/ensemble/ensemble.h
	spinnaker_components/ensemble/neuron_lif.h
	spinnaker_components/filter/filter_main.c
	spinnaker_components/value_sink/value_sink.c
	tests/operators/test_lif.py
	tests/regions/test_region_recording.py

neworderofjamie committed:
More tests

neworderofjamie committed:
Binaries

neworderofjamie committed:
Scipy dependency

neworderofjamie committed:
Actually, for correctness, binned_statistic was present in 0.11 of SciPy

neworderofjamie committed:
Attempt to install scipy from apt-get

neworderofjamie committed:
Attempt to install scipy from apt-get (reverted from commit 9db5dfcc386a893d4102a47ae62de575e7d8ff23)

neworderofjamie committed:
Attempt alternative means of installing scipy

neworderofjamie committed:
lapack

neworderofjamie committed:
Umm, dependency-resolution much!?!

neworderofjamie committed:
Already bored of build times - made scipy optional and imported it where it's used

neworderofjamie committed:
PEP8 and fixes

neworderofjamie committed:
pepepep

neworderofjamie committed:
Python 3

mundya commented:
Manually merged into master.
